src/components/Field/Field.tsx

```diff
@@ -0,0 +1,86 @@
+import React, { ComponentType, forwardRef } from "react";
+import { FieldPathValue, FieldPath, FieldValues, RegisterOptions } from "react-hook-form";
+import { useBasisField } from "../../hooks/useBasisForm/useBasisField";
+import mergeRefs from "../../utils/mergeRefs";
+
+export type Component<Props> =
+  | ComponentType<Props>
+  | React.ForwardRefExoticComponent<Props & React.RefAttributes<any>>
+  | (React.ForwardRefExoticComponent<Props & React.RefAttributes<any>> &
+      Record<string, any>)
+  | (React.Component<Props> & Record<string, any>);
+
+type ValueProps<Value> = {
+  value?: Value;
+};
+
+type FieldProps<
+  TFieldValues extends FieldValues = FieldValues,
+  Name extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
+  Props = ValueProps<FieldPathValue<TFieldValues, Name>>
+> =
+  // check if value of component passed from `as` matches the
+  Props extends ValueProps<FieldPathValue<TFieldValues, Name>>
+    ? // Infer the type of Error expected by the validate function
+      {
+        name: Name;
+        defaultValue?: FieldPathValue<TFieldValues, Name>;
+        rules?: RegisterOptions<TFieldValues, Name>;
+        // Have to use Component<Props> to allow components with custom properties and forward refs
+        as: Component<Props>;
+      } & Props
+    : // show nicer error message for component mismatch
+    "Component in `as=` expects a different value than the one provided by `name=`";
+
+export type FieldComponent<TFieldValues extends FieldValues = FieldValues> = <
+  Name extends FieldPath<TFieldValues>,
+  P = ValueProps<FieldPathValue<TFieldValues, Name>>
+>(
+  props: FieldProps<TFieldValues, Name, P>
+) => any;
+
+
+export const Field = forwardRef(<
+    TFieldValues extends FieldValues,
+    Name extends FieldPath<TFieldValues>,
+    Props extends ValueProps<FieldPathValue<TFieldValues, Name>>
+  >(
+    props: FieldProps<TFieldValues, Name, Props>,
+    ref: any
+  ) => {
+    const {
+      name,
+      defaultValue,
+      rules,
+      as: Comp,
+      ...componentProps
+    } = props;
+
+    const Component = Comp as ComponentType<any>;
+
+    const {
+      onChange,
+      onBlur,
+      ref: formComponentRef,
+      error,
+      ...childProps
+    } = useBasisField<TFieldValues, Name>({
+      name,
+      componentDisplayName: Component.displayName,
+      componentProps: componentProps,
```

Comment:
`componentProps,` should be enough

Reply:
They are not enough, as they don't give the display name of the component in the props, that is a static property that exists only on the component reference.